packages/public/server/src/module/Api/src/Controller/AbstractApiController.php

```diff
@@ -53,4 +58,27 @@ protected function assertAuthorization()
             return new JsonModel(['reason' => 'Invalid authorization header']);
         }
     }
+
+    protected function getRequestBody(array $validators): array
+    {
+        $data = Json::decode(
+            $this->getRequest()->getContent(),
+            Json::TYPE_ARRAY
+        );
+        if (!is_array($data)) {
+            throw new \TypeError();
+        }
+        $keys = Utils::array_union(array_keys($data), array_keys($validators));
+        $isValid = Utils::array_every(function ($key) use ($data, $validators) {
+            return isset($validators[$key]) &&
+                isset($data[$key]) &&
+                $validators[$key]($data[$key]);
+        }, $keys);
+
+        if ($isValid) {
+            return $data;
+        } else {
+            throw new \TypeError();
+        }
+    }
```

Comment:
I haven't found a better way to implement it with a form. I would suggest to do this in a follow up issue.

Reply:
Shouldn't this already be covered in your code anyways, e.g.:
```php
$data = $this->getRequestBody([
    'content' => 'is_string',
    'userId' => 'is_int',
    'threadId' => 'is_int',
]);

// ...

/** @var CommentForm */
$form = $this->getServiceLocator()->get(CommentForm::class);
// The form will validate the given data.
$form->setData([
    'parent' => $uuid,
    'author' => $user,
    'instance' => $uuid->getInstance(),
    'content' => $data['content'],
    'csrf' => CsrfTokenContainer::getToken(),
    'subscription' => [
        'subscribe' => true,
        'mailman' => true,
    ],
]);

```

Reply:
Unfortunately not. When `$data["title"]` does not exists PHP does not raise an exception but raises an uncatchable `undefined index` error. So a `isset($data["content"]) ? $data["content"] : ...` is needed. Also I haven't found an easy way so that the form makes a check for the `$title` whether it is a string or not.

Reply:
Another problem is `$data["userId"]` -> With `(int) $data["userId"] ?? -1` a non empty string for `$data["userId"]` is converted into 1 and then aeneas is found in the database. So I ended with something like `$userId = isset($data["userId") && is_int($data["userId"]) ? $data["userId"] : -1`. I haven't found a way to avoid such a utility function.

Reply:
Well, whatever. Let's leave it like that then since we'll move it to the database layer soon, anyway.